pkg/limits/stream_metadata.go

```diff
@@ -0,0 +1,232 @@
+		}
+	}
+}
+func (s *streamMetadata) Store(tenant string, partitionID int32, streamHash, recTotalSize uint64, recordTime, bucketStart, bucketCutOff int64) {
+	i := s.getStripeIdx(tenant)
+
+	s.locks[i].Lock()
+	defer s.locks[i].Unlock()
+
+	// Initialize stripe map if it doesn't exist
+	if s.stripes[i] == nil {
+		s.stripes[i] = make(map[string]map[int32][]Stream)
+	}
+
+	// Initialize tenant map if it doesn't exist
+	if _, ok := s.stripes[i][tenant]; !ok {
+		s.stripes[i][tenant] = make(map[int32][]Stream)
+	}
+
+	// Initialize partition map if it doesn't exist
+	if s.stripes[i][tenant][partitionID] == nil {
+		s.stripes[i][tenant][partitionID] = make([]Stream, 0)
+	}
+
+	for j, stream := range s.stripes[i][tenant][partitionID] {
+		if stream.Hash == streamHash {
+			// Update total size
+			totalSize := stream.TotalSize + recTotalSize
+
+			// Update or add size for the current bucket
+			updated := false
+			sb := make([]RateBucket, 0, len(stream.RateBuckets)+1)
+
+			// Only keep buckets within the rate window and update the current bucket
+			for _, bucket := range stream.RateBuckets {
+				// Clean up buckets outside the rate window
+				if bucket.Timestamp < bucketCutOff {
+					continue
+				}
+
+				if bucket.Timestamp == bucketStart {
+					// Update existing bucket
+					sb = append(sb, RateBucket{
+						Timestamp: bucketStart,
+						Size:      bucket.Size + recTotalSize,
+					})
+					updated = true
+				} else {
+					// Keep other buckets within the rate window as is
+					sb = append(sb, bucket)
+				}
+			}
+
+			// Add new bucket if it wasn't updated
+			if !updated {
+				sb = append(sb, RateBucket{
+					Timestamp: bucketStart,
+					Size:      recTotalSize,
+				})
+			}
+
+			s.stripes[i][tenant][partitionID][j] = Stream{
+				Hash:        stream.Hash,
+				LastSeenAt:  recordTime,
+				TotalSize:   totalSize,
+				RateBuckets: sb,
+			}
+			return
+		}
+	}
+
+	// Create new stream metadata with the initial interval
+	s.stripes[i][tenant][partitionID] = append(s.stripes[i][tenant][partitionID], Stream{
+		Hash:        streamHash,
+		LastSeenAt:  recordTime,
+		TotalSize:   recTotalSize,
+		RateBuckets: []RateBucket{{Timestamp: bucketStart, Size: recTotalSize}},
+	})
+}
+
+func (s *streamMetadata) Evict(cutoff int64) map[string]int {
+	evicted := make(map[string]int)
+
+	for i := range s.locks {
+		s.locks[i].Lock()
+
+		for tenant, streams := range s.stripes[i] {
+			for partitionID, partition := range streams {
+				activeStreams := make([]Stream, 0)
+
+				for _, stream := range partition {
+					if stream.LastSeenAt >= cutoff {
+						activeStreams = append(activeStreams, stream)
+					} else {
+						evicted[tenant]++
+					}
+				}
+
+				s.stripes[i][tenant][partitionID] = activeStreams
+			}
+		}
+		s.locks[i].Unlock()
+	}
+
+	return evicted
+}
+
+func (s *streamMetadata) EvictPartitions(partitions []int32) {
+	for i := range s.locks {
+		s.locks[i].Lock()
+
+		for _, id := range partitions {
```

Comment:
Would it be faster to invert the order of this? Instead of accessing `s.stripes[i]` `len(partition)` times, instead just access it once:

```go
for tenant, partitions := range s.stripes[i] {
	for _, deleteId := range partitions {
		delete(partitions, deleteId)
	}
	if len(partitions) == 0 {
		delete(s.stripes[i], tenant)
	}
}
```

This should be faster as then we don't suffer loads of cache misses, which means holding the lock for less time.